editor/gui/editor_scroll_box.cpp

```diff
@@ -0,0 +1,277 @@
+			second_button->set_tooltip_text(TTRC("Scroll Left\nHold Ctrl to scroll to the begin.\nHold Shift to scroll one page."));
+			first_button->set_button_icon(theme_cache.arrow_right);
+			first_button->set_accessibility_name(TTRC("Scroll Right"));
+			first_button->set_tooltip_text(TTRC("Scroll Right\nHold Ctrl to scroll to the end.\nHold Shift to scroll one page."));
+		} else {
+			first_button->set_button_icon(theme_cache.arrow_left);
+			first_button->set_accessibility_name(TTRC("Scroll Left"));
+			first_button->set_tooltip_text(TTRC("Scroll Left\nHold Ctrl to scroll to the begin.\nHold Shift to scroll one page."));
+			second_button->set_button_icon(theme_cache.arrow_right);
+			second_button->set_accessibility_name(TTRC("Scroll Right"));
+			second_button->set_tooltip_text(TTRC("Scroll Right\nHold Ctrl to scroll to the end.\nHold Shift to scroll one page."));
+		}
+	} else {
+		first_button->set_button_icon(theme_cache.arrow_up);
+		first_button->set_accessibility_name(TTRC("Scroll Up"));
+		first_button->set_tooltip_text(TTRC("Scroll Up\nHold Ctrl to scroll to the begin.\nHold Shift to scroll one page."));
+		second_button->set_button_icon(theme_cache.arrow_down);
+		second_button->set_accessibility_name(TTRC("Scroll Down"));
+		second_button->set_tooltip_text(TTRC("Scroll Down\nHold Ctrl to scroll to the end.\nHold Shift to scroll one page."));
+	}
+}
+
+void EditorScrollBox::_update_disabled_buttons() {
+	ScrollBar *scroll_bar = is_vertical() ? (ScrollBar *)scroll_container->get_v_scroll_bar() : (ScrollBar *)scroll_container->get_h_scroll_bar();
+	if (!is_vertical() && is_layout_rtl()) {
+		first_button->set_disabled(scroll_bar->get_value() + scroll_bar->get_page() == scroll_bar->get_max());
+		second_button->set_disabled(scroll_bar->get_value() == 0);
+	} else {
+		first_button->set_disabled(scroll_bar->get_value() == 0);
+		second_button->set_disabled(scroll_bar->get_value() + scroll_bar->get_page() == scroll_bar->get_max());
+	}
+}
+
+bool EditorScrollBox::_set(const StringName &p_name, const Variant &p_value) {
+	if (p_name == "vertical") {
+		if ((bool)p_value == is_vertical()) {
+			return true;
+		}
+		set_vertical((bool)p_value);
+		if (is_vertical() != (bool)p_value) {
+			return true;
+		}
+
+		// switch the size flags
+		if (is_vertical()) {
+			set_v_size_flags(get_h_size_flags());
+			set_h_size_flags(SIZE_FILL);
+		} else {
+			set_h_size_flags(get_v_size_flags());
+			set_v_size_flags(SIZE_FILL);
+		}
+
+		_update_buttons_icon_and_tooltip();
+		_update_buttons();
+		_update_scroll_container();
+
+		emit_signal(SNAME("vertical_changed"), p_value);
+		return true;
+	}
+	return false;
+}
+
+void EditorScrollBox::_validate_property(PropertyInfo &p_property) const {
+	if (is_fixed && p_property.name == "vertical") {
+		p_property.usage = PROPERTY_USAGE_NONE;
+	}
+}
+
+void EditorScrollBox::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_TRANSLATION_CHANGED:
```

Comment:
```suggestion
		case NOTIFICATION_ACCESSIBILITY_UPDATE: {
			RID ae = get_accessibility_element();
			ERR_FAIL_COND(ae.is_null());

			DisplayServer::get_singleton()->accessibility_update_set_role(ae, DisplayServer::AccessibilityRole::ROLE_SCROLL_VIEW);
			DisplayServer::get_singleton()->accessibility_update_add_action(ae, DisplayServer::AccessibilityAction::ACTION_SCROLL_LEFT, callable_mp(this, &EditorScrollBox::_accessibility_action_scroll_left));
			DisplayServer::get_singleton()->accessibility_update_add_action(ae, DisplayServer::AccessibilityAction::ACTION_SCROLL_RIGHT, callable_mp(this, &EditorScrollBox::_accessibility_action_scroll_right));
		} break;
		case NOTIFICATION_TRANSLATION_CHANGED:
```
Mark it as scrollable and add action callbacks.

Reply:
Will use `ACTION_SCROLL_BACKWARD` and `ACTION_SCROLL_FORWARD` instead of `ACTION_SCROLL_LEFT` and `ACTION_SCROLL_RIGHT` since it also has a vertical layout.

Reply:
`ACTION_SCROLL_BACKWARD` and `ACTION_SCROLL_FORWARD` were removed in latest version of accesskit, and only kept for compatibility, they are aliases for `UP` and `DOWN`.

Reply:
So for vertical it should be `if` with LEFT/RIGHT and UP/DOWN branches.

Reply:
Can we instead use those four action and return early when the action is not supported for vertical or horizontal modes ?

```c++
	void _accessibility_action_scroll_up(const Variant &p_data);
	void _accessibility_action_scroll_down(const Variant &p_data);
	void _accessibility_action_scroll_left(const Variant &p_data);
	void _accessibility_action_scroll_right(const Variant &p_data);
```

Reply:
I think it should not have all 4 actions registered at the same time, since it will show up as if it has both scroll directions, while there's only one.

Reply:
> I think it should not have all 4 actions registered at the same time, since it will show up as if it has both scroll directions, while there's only one.

We can register the `EditorScrollBox` as an abstract class and we only add two actions if it's vertical or horizontal.

Reply:
Yes, it looks fine.